### ISSUE TITLE
`createTexture` fails with `dimension: '3d', format: 'r32float'`

### ISSUE THREAD

**Issue by xinaesthete** [visgl/luma.gl#2115]
I'm trying to update viv to the latest deck&luma and have an issue getting 3D textures to work.

https://github.com/hms-dbmi/viv/pull/805#issuecomment-2186381214

In xr-3d-layer.js, we call this (I believe that it is now not necessary to pass anything other than the format as a string, ie. 'r32float'):

    const attrs = getRenderingAttrs(); // attrs.format = 'r32float'
    const texture = this.context.device.createTexture({
      width,
      height,
      depth,
      dimension: '3d',
      data: attrs.cast?.(data) ?? data,
      format: attrs.format,
      mipmaps: false,
      sampler: {
        minFilter: 'linear',
        magFilterFilter: 'linear',
        addressModeU: 'clamp-to-edge',
        addressModeV: 'clamp-to-edge',
        addressModeW: 'clamp-to-edge'
      }
    });

luma.gl in the process of initialising the texture ends up calling `convertTextureFormatToGL(format)` in webgl-texture `setImageData3D()`, having already converted from string format to GL constant (33326). So it throws an "Unsupported texture format error". I briefly tried to patch my local version of luma to see how far it'd get if I overrode that, but didn't manage to apply that correctly.

To reproduce, you can clone my PR and use a sample with 3D data such as https://viv-demo.storage.googleapis.com/brain.pyramid.ome.tif, press the "show volumetric rendering" button, and select a resolution (probably the lowest, for testing).

**Comment by xinaesthete** [comment]
I've cloned luma.gl and modified the texture-3d example to use `'r32float'` format, which doesn't reproduce the error. I'm a little confused currently about where `setImageData3D` even is, hopefully it'll make a bit more sense when I get back to this with a clearer head.

**Comment by xinaesthete** [comment]
I had hoped that the new release would fix this, but it doesn't seem so.

Not sure why - I recalled the example in the main branch working some time ago.

Not having working texture-3d is a bit of a blocker, unfortunately.

https://github.com/xinaesthete/luma-texture3d

**Comment by ibgreen** [comment]
Hi, apologies, WebGL Texture support is undergoing a major refactor in luma.gl 9.1 to streamline and align with the WebGPU implementation.

I'll admit that I have been struggling with the texture 3d support for some reason I haven't been able to get the example working again on the branch in spite spending quite some time on it. I'm getting ready to give it another go soon.

> . I'm a little confused currently about where setImageData3D

Yes, one of the problems is that these texture setter methods need to be replaced to match the way WebGPU handles texture data writes (through command buffers), so there will most likely be some breaking API changes.

**Comment by xinaesthete** [comment]
> Hi, apologies, WebGL Texture support is undergoing a major refactor in luma.gl 9.1 to streamline and align with the WebGPU implementation.
> 
> I'll admit that I have been struggling with the texture 3d support for some reason I haven't been able to get the example working again on the branch in spite spending quite some time on it. I'm getting ready to give it another go soon.
> 
> > . I'm a little confused currently about where setImageData3D
> 
> Yes, one of the problems is that these texture setter methods need to be replaced to match the way WebGPU handles texture data writes (through command buffers), so there will most likely be some breaking API changes.

Thanks very much for the response - it sounds like something that's probably better left to you at this point, although I'm a little bit tempted to possibly have a go at a PR if that might be any help.

Good luck!

**Comment by ibgreen** [comment]
@xinaesthete If you are experienced with 3d textures and willing to try your hand at debugging the texture-3d example, a fresh pair of eyes could be very helpful. 

As far as I have been able to ascertain, the texture is created with the right values and data is uploaded (at least the first "slice", webgl doesn't appear to read back all the depth slices).

But sampling from shader just doesn't seem to work.
- sampler setting issue?
- mipmaps?
- ...?

```
git co master
yarn bootstrap
cd examples/api/texture-3d
yarn
yarn start
```

**Comment by xinaesthete** [comment]
ok @ibgreen I'll give it a shot - not all that highly confident, but very keen to get it working and help where I can.

FWIW when I checked out the repo before, thinking that I might make a PR if I got anywhere useful, I found that I needed a slight bit of fiddling to get `yarn bootstrap` to work - I'm not very used to yarn and not entirely sure why I had the issue. I eventually found that [`yarn set version berry` helped](https://github.com/xinaesthete/luma.gl/commit/da207418e0a267683151435ab88c78a72832bc13).

**Comment by xinaesthete** [comment]
I'm in BST and I'm going to start having a quick look now and see how I get on, I'll be leaving work moderately soon, though.

**Comment by xinaesthete** [comment]
Hmmm...

So far I've tried in my previous `tex3d` branch - where I hadn't made any relevant changes, but it was based on a version of luma that does work. The current version I can reproduce the error. I've tried looking at the state with spector.js, but from what I can see so far they both look pretty identical.

![luma-debug](https://github.com/user-attachments/assets/e8719e06-8fb2-4f95-8565-72c98238f883)

So I guess the next step is to check what relevant parts of the code have changed since when I forked.

**Comment by xinaesthete** [comment]
Update: the example does work on current master when it's changed to format `'r32float'` (and corresponding `Float32Array`).

One hypothesis is that it's just to do with whether the data is treated as normalized.

**Comment by xinaesthete** [comment]
Yep... it looks like it's working with `fragColor.rgb = fragColor.rgb * 256.0;` in the shader.

Hopefully I'm not mistaken in my analysis - seems like one of those times when a second pair of eyes is indeed helpful!

**Comment by xinaesthete** [comment]
@ibgreen any idea when it'll make it to a release version? Obviously you know better what else is in flux etc, but it'd definitely be great if it could.

That said - I tried adapting the sample to WebGPU

```js
  // luma.registerAdapters([webgl2Adapter]);
  luma.registerDevices([WebGPUDevice]);
  makeAnimationLoop(AnimationLoopTemplate, {type: "webgpu"}).start();
```

and am getting errors in `webgpu-texture.js` `setData()` with bad arguments to `new ImageData()` - both `this.width` and `this.height` are `1`, which they probably shouldn't be... so now I'm here and debugging, I'll look into that and make a PR if it makes sense to. Currently I still have issues with my IDE setup, showing errors like `Cannot find module '@luma.gl/engine' or its corresponding type declarations.`, but at least the code is running and editable.

**Comment by xinaesthete** [comment]
The trouble with

```sh
cd examples/api/texture-3d
yarn
yarn start
```

is that it doesn't rebuild the actual library code... I'm sure I'll figure that out.

**Comment by xinaesthete** [comment]
Obviously there's a fair bit more work to get it working with WebGPU - `setTexture3DData` et al not implemented, and `Error: GLSL shaders are not supported in WebGPU`.

Is it out-of-scope to try to transpile `#version 300 es` shaders? I think I'd vaguely thought that this may be a feature - but I'm not asserting that it *should* be.

> The luma.gl v9 API is designed to provide portable WebGPU / WebGL 2 API, ground-up TypeScript support, and strong support for working with both WGSL and GLSL shaders.

I notice that e.g. in [babylon.js](https://doc.babylonjs.com/setup/support/webGPU/webGPUWGSL)

> Currently, all shaders used by Babylon.js are written in [GLSL](https://www.khronos.org/opengl/wiki/OpenGL_Shading_Language) and are converted to [WGSL](https://gpuweb.github.io/gpuweb/wgsl/) (the only shader language that WebGPU knows about) by some special tools.
>
> So, even in WebGPU, if you use a CustomMaterial or a PBRCustomMaterial to inject some custom shader code, you must write it in [GLSL](https://www.khronos.org/opengl/wiki/OpenGL_Shading_Language).

and I think I thought I'd interpreted

> GLSL 1.00 is no longer supported. GLSL shaders need to be ported to GLSL 3.00.

as meaning that something somewhat similar would be done here - but again, if that's not the plan I'm not about to suggest otherwise.

Anyway, veering somewhat off-topic for the actual issue at hand.

Best,
Peter

**Comment by ibgreen** [comment]
> Is it out-of-scope to try to transpile #version 300 es shaders? I think I'd vaguely thought that this may be a feature - but I'm not asserting that it should be.

At the moment we don't have any transpilation, shaders need to be written in both GLSL (.vs, .fs props) and WGSL (.source prop). 

luma.gl v9 is WebGPU first so a possible long term goal would be to provide WGLS -> GLSL transpilation.

**Comment by ibgreen** [comment]
> setTexture3DData et al not implemented

A problem is that in WebGPU there are no primitives to upload from CPU arrays to textures (for some reason there is such an API for CPU Images). 

The required approach is to create a CommandQueue, upload to a Buffer and copy buffer to texture. Working on a new API for this, but can be done with raw webgpu commands as well.

**Comment by xinaesthete** [comment]
> At the moment we don't have any transpilation, shaders need to be written in both GLSL (.vs, .fs props) and WGSL (.source prop).
> 
> luma.gl v9 is WebGPU first so a possible long term goal would be to provide WGLS -> GLSL transpilation.

I suspect the source transpilation itself wouldn't be too bad, but attempting to support all the appropriate mappings of uniforms/resources etc might be a hassle - hopefully not such a massive burden for users to port their shaders.

**Comment by xinaesthete** [comment]
Any idea when this'll make it into a release? 9.0.17 still has the original issue.

I think it just needs a change to [this line in webgl-texture.ts](https://github.com/visgl/luma.gl/blob/de7ed0433a0815ea9bb36b7e917c6b0c690e960a/modules/webgl/src/adapter/resources/webgl-texture.ts#L314), so that instead of `format: glFormat` you use `format: props.format`.

I couldn't get the texture-3d example to run properly on the `9.0-release` branch because of other issues - I found that all the examples I tried needed a `{device}` passed to `makeAnimationLoop` or there's another error

> Cannot access 'WebGPUDevice' before initialization

The texture-3d example also had another error with the position buffer not having layout attribute or something... I'll look into that now. However, with the above change, it no longer throws an error while creating the texture, and the `hello-cube` example which goes through the same `setImageData` code-path still works with that change.

I would really be very grateful for a release with this issue fixed ASAP - I'm finding myself wanting to do quite a few things that'll either benefit from new features / community layers in deck.gl, or will need to be re-written for changes in luma.gl API - and would also involve figuring out how to do them the old way based on old versions of the docs, when I'm fairly convinced the new ways are better.
